### ISSUE TITLE
Questions about the paper

### ISSUE THREAD

**Issue by WOWNICE** [ZHKKKe/DualStudent#4]
Hello, your paper is indeed exciting and insightful, while after I finished reading, I still have some questions unsolved. 
I will appreciate it if you could spend some time answering the following:

1. You proved that the teacher and the student converge to the same point (eventually), but why this convergence is bad anyway, and why can't they converge to a rather good solution? 

2. Secondly, I wonder why the idea of "stable samples" works. Sometimes I think we don't care much about them: for example in active learning people tend to know those near the decision boundary, which are "unstable samples", because they carry more information than stable ones. In your paper, with all due respect and let me point it out, you speculated that "applying the consistency constraint directly will cause them to collapse into each other by exchanging the wrong knowledge", which is not supported by any evidence, and I wonder why the proposed "stable points", on which the consistency constraint is applied, could help solve the collapsing problem you mentioned. Do you have any other derivation on that? 

3. Finally, you spent rather short term of discourse to do the ablation study and I'm curious whether there is numerical performance when you not "training with stable samples"? 

Thanks for your time again! In general I think the overall idea is very insightful!

**Comment by ZHKKKe** [comment]
Hi! Thanks for your attention! Since the answer below may be a bit long, I bold some sentences as **TL;DR**.
<br>

--------------------------------------------------------------------------
### **Question 1**
The statement *“the teacher and the student converge to the same point”* does not mean that they will converge to a bad solution. On the contrary, since semi-supervised learning is effective, both models (although they are coupled) should converge to a better solution than the model trained by the labeled data only. 

The problem is, in the late stage of training, **the two coupled models have very similar predictions for the same unlabeled data**. In this case, **the consistency loss between the two models could not learn more from the unlabeled data, i.e., the performance bottleneck mentioned in the paper**. Please refer to Fig. 7 (left) in the paper, MT (blue curve) converges faster than DS (yellow curve) in the early stage, but MT has little performance improvement after model coupling.  

### **Question 2**
In both semi-supervised and active learning (or other similar tasks), data points near the decision boundary should carry more information. However, **the pseudo labels constructed from these data points are also more unreliable, i.e., more likely to be wrong**. In our case, directly using the prediction of one student model as a pseudo label of another student model will **introduce too much noise (too many wrong pseudo labels)** and cause the training to collapse. Therefore, **we define stable sample to obtain relatively reliable pseudo labels**. As shown in Fig. 6 in the paper, the model’s predictions for stable samples are more likely to be correct, which means the model’s predictions for them can be regard as more reliable pseudo labels.

Besides, **stable samples explore data points close to the decision boundary to a certain degree**. For a given stable sample $x$, the definition of the stable sample dose not constrain the model’s prediction probabilities of $\bar(x)$ (near points of $x$) to be higher than the threshold. Therefore, $f(\bar{x})$ may be close to the decision boundary.

For the statement *“applying the consistency constraint directly will cause them to collapse into each other by exchanging the wrong knowledge”*, if I remember correctly, **[\*** we have trained two independent student models by using the consistency loss, but the final performance was very poor **\*]**. I will re-validate this statement and update the result to you when I am free. I think the collapse is caused by the too many wrong pseudo labels mentioned above.

### **Question 3**
This question is interesting. **Since the proposed stability constraint is based on the stable samples, we did not conduct ablation for the stable sample**. If using all the samples for training, how to judge which student model has a more reliable prediction for an unlabeled data? Only use the prediction probability (it may not enough)? This may be a direction for future work.

In addition, if the conclusion **[\*...\*]** in **Question 2** is correct, **the “CS” experiment in Table 1 (in the paper) can show that the stabile sample can solve the model collapse** mentioned in **Question 2**, i.e., the stable samples are effective.

--------------------------------------------------------------------------

<br>
I hope the above answers can solve your confusions. If you have other questions, please feel free to ask below.  :)

**Comment by WOWNICE** [comment]
OMG thank you so much for such an exhaustive reply! I didn't expect you would do such a nice thing! About the first question, since I'm new to the field of SSL, correct me if I'm wrong: 

1. Originally there is only one model in SSL, and the consistency constraint is applied to the model itself, which utilizes the information in the unlabelled data. 
2. The reason why we use the T-S framework is that the consistency loss is too fluctuating, and averaging the model weights would help alleviate this. 

So I think based on the two claims above, we can say that **the EMA updating method proposed in Mean Teacher is not the component that helps the model to exploit the unlabelled data (but the consistency constraint)**. And when the student and teacher converge to each other, it means that the student network is stable enough (because it converges). At this time I speculate (without any evidence XD) consistency loss is still contributing to learning from the unlabelled data. So I think that EMA, as a smoothing method, doesn't really have a "bottleneck problem". 

As for the remaining two questions, I am still thinking about them, thanks for your reply again!

**Comment by ZHKKKe** [comment]
You are welcome.

### For Question 1:

Yes, Mean Teacher (MT) uses EMA to ensemble a more reliable teacher to alleviate the fluctuation of the consistency loss. Notice that EMA does not cause the "bottleneck" directly.

**The coupled weights of two models cause the "bottleneck", and EMA is the reason for coupling. Such coupling happens in all existing Teacher-Student structures**, such as Pi Model, Temporal Model, and MT. Actually, we can treat (1) Pi Model as EMA with alpha=0, (2) Temporal Model as EMA of the pseudo-labels, and (3) MT as EMA of the models. In our paper, we just used MT as an example. 

Let's consider the most extreme case of MT. When the weights of the teacher model are the same as the student model (which will eventually occur in MT due to EMA), the two models' predictions are exactly the same (assume the perturbations are disabled). It means that the consistency loss can learn nothing from unlabeled data. Of course, in practice, the perturbations are applied to alleviate this problem, but it is not enough to decouple the two models. This is also why the perturbations are important to the consistency loss, without the perturbation, Pi Model and the late stage of MT will learn nothing from the unlabeled data. 

**In summary, the coupled student and teacher have very similar predictions, which makes the consistency loss cannot learn from unlabeled data effectively. This is the "bottleneck".** Dual Student's idea is to break the "bottleneck" by decoupling the two models as much as possible.
